src/Microsoft.AspNet.Server.Kestrel/Http/SocketOutput.cs

```diff
@@ -80,23 +78,25 @@ public SocketOutput(KestrelThread thread, UvStreamHandle socket, long connection
                 {
                     _nextWriteContext.SocketDisconnect = true;
                 }
+
                 // Complete the write task immediately if all previous write tasks have been completed,
                 // the buffers haven't grown too large, and the last write to the socket succeeded.
-                triggerCallbackNow = _lastWriteError == null &&
-                                     _callbacksPending.Count == 0 &&
-                                     _numBytesPreCompleted + buffer.Count <= _maxBytesPreCompleted;
-                if (triggerCallbackNow)
+                if (_lastWriteError == null &&
+                    _tasksPending.Count == 0 &&
+                    _numBytesPreCompleted + buffer.Count <= _maxBytesPreCompleted)
                 {
                     _numBytesPreCompleted += buffer.Count;
                 }
+                else if (immediate)
+                {
+                    // immediate write, which is not eligable for instant completion above
+                    tcs = new TaskCompletionSource<object>(buffer.Count);
+                    _tasksPending.Enqueue(tcs);
+                }
                 else
                 {
-                    _callbacksPending.Enqueue(new CallbackContext
-                    {
-                        Callback = callback,
-                        State = state,
-                        BytesToWrite = buffer.Count
-                    });
+                    // immediate==false calls always return complete tasks, because there is guaranteed
+                    // to be a subsequent immediate==true call which will go down one of the previous code-paths
```

Comment:
I hear you. I wish @davidfowl would ease up a little on requiring travis to pass given the current state of Kestrel on Mono. For now, it's just motivation for me to fix the timeouts :persevere: